src/bluesearch/database/article.py

```diff
@@ -227,24 +215,78 @@ def get_paragraphs_sections_mapping(self) -> Dict[Element, str]:
 
         return mapping
 
-    @staticmethod
-    def text_content(element: Optional[Element]) -> str:
-        """Extract all text of an element and of its descendants (at any depth).
+    def _inner_text(self, element: Element) -> str:
+        """Convert all inner text and sub-elements to one string.
+
+        In short, we collect all the inner text while also converting all
+        sub-elements that we encounter to strings using ``self._element_to_str``.
+        All escaped HTML in the raw text is unescaped.
+
+        For example, if schematically the element is given by
+
+            element = "<p>I <bold>like</bold> python &amp; ice cream.<p>"
+
+        then ``_inner_text(element)`` would give
+
+            "I like python & ice cream."
+
+        provided that "<bold>like</bold>" is resolved to "like" by the
+        ``self._element_to_str`` method.
+
+        Parameters
+        ----------
+        element
+            The input XML element.
+
+        Returns
+        -------
+        str
+            The inner text and sub-elements converted to one single string.
+        """
+        text_parts = [html.unescape(element.text or "")]
+        for sub_element in element:
+            # recursively parse the sub-element
+            text_parts.append(self._element_to_str(sub_element))
+            # don't forget the text after the sub-element
+            text_parts.append(html.unescape(sub_element.tail or ""))
+        return "".join(text_parts).strip()
+
+    def _element_to_str(self, element: Element | None) -> str:
+        """Convert an element and all its contents to a string.
```

Comment:
Hm, interesting, I guess we should talk about it. Because the motivation was exactly to make the NLP task easier. Isn't it common to do this kind of replacement in NLP?

Reply:
Great point, @pafonta! I guess this PR gives us the possibility to easily change (via hardcoding) the placeholders. I guess it is debatable whether using a placeholder harms the downstream NLP task.

If we take the wordpiece tokenizer and assume we are parsing a latex formula. On one hand, the tokenizer will tokenize it without major issues and theoretically the tokens will hold some useful meaning `=, \frac{}{}, ...`. However, the main problem in the past was that the `# of tokens` after tokenization that can get really big. IMO removing/replacing this formula at parse time is probably the right moment to deal with this problem. Mostly because the XML tag precisely tells you where they start and where they end.

Reply:
> Isn't it common to do this kind of replacement in NLP?

Not for language modeling (see next comment below).

> I guess it is debatable whether using a placeholder harms the downstream NLP task.

Placeholders harms language modeling, especially when fine-tuning, like we do. Indeed, placeholders are disrupting the learned language syntax and semantic.

We use masked language modeling. This means that the model is trained by predicting a token depending of its surrounding tokens. If such tokens are placeholders, we mess the training. Besides, using placeholders means that the trained language model needs to be retrained each time a placeholder change or is added.

Reply:
This being said, we could merge this PR which adds placeholders but add a follow-up issue about either:
- translating the content into a string,
- completely removing the content from the text but storing it separately for later use.

Reply:
> Isn't it common to do this kind of replacement in NLP?

> Not for language modeling (see next comment below).

I think both of these approaches sound reasonable. It would be interesting to see whether/how other people treated emails, urls etc. in the context of NLP problems using transformer-based models. We can have a quick look and then decide what to do accordingly.

➡️ **What about opening a dedicated issue and briefly discussing a couple of examples that we can find online there, and then implement a chosen strategy?**

---

Here are my two-cents on the subject. I know very little so I would be glad to change my mind if presented with info going in the opposite direction.
1. I think this process of replacing sensitive/meaningless info with something else is sometimes referred to as "pseudonymization".
2. I think have seen this being done also with BERT models. See e.g. [this project](https://github.com/digitalepidemiologylab/covid-twitter-bert/blob/c87912b409659f40018e839c4124be5ae2486713/preprocess/prepare_pretrain_data.py#L140-L144), where they used "a transformer-based model pretrained on a large corpus of Twitter messages on the topic of COVID-19".

Reply:
Created #447